cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,805 @@
+      detail::is_first_in_run_t<decltype(type_and_gpu_id_pair_begin)>{type_and_gpu_id_pair_begin});
+
+    auto unique_pairs_buffer = cugraph::allocate_dataframe_buffer<
+      typename thrust::iterator_traits<decltype(type_and_gpu_id_pair_begin)>::value_type>(
+      nr_unique_paris, handle.get_stream());
+
+    rmm::device_uvector<edge_t> unique_pair_counts(nr_unique_paris, handle.get_stream());
+
+    thrust::reduce_by_key(handle.get_thrust_policy(),
+                          type_and_gpu_id_pair_begin,
+                          type_and_gpu_id_pair_begin + edge_types.size(),
+                          thrust::make_constant_iterator(size_t{1}),
+                          cugraph::get_dataframe_buffer_begin(unique_pairs_buffer),
+                          unique_pair_counts.begin());
+
+    edge_types.resize(0, handle.get_stream());
+    gpu_ids.resize(0, handle.get_stream());
+
+    std::forward_as_tuple(
+      std::tie(
+        std::get<0>(unique_pairs_buffer), std::get<1>(unique_pairs_buffer), unique_pair_counts),
+      std::ignore) =
+      cugraph::groupby_gpu_id_and_shuffle_values(
+        handle.get_comms(),
+        thrust::make_zip_iterator(thrust::make_tuple(std::get<0>(unique_pairs_buffer).begin(),
+                                                     std::get<1>(unique_pairs_buffer).begin(),
+                                                     unique_pair_counts.begin())),
+        thrust::make_zip_iterator(thrust::make_tuple(std::get<0>(unique_pairs_buffer).end(),
+                                                     std::get<1>(unique_pairs_buffer).end(),
+                                                     unique_pair_counts.end())),
+        [] __device__(auto val) { return thrust::get<1>(val); },
+        handle.get_stream());
+
+    //
+    // Count local #elments for all the types mapped to this GPU
+    //
+
+    thrust::sort_by_key(handle.get_thrust_policy(),
+                        std::get<0>(unique_pairs_buffer).begin(),
+                        std::get<0>(unique_pairs_buffer).end(),
+                        unique_pair_counts.begin());
+
+    auto nr_unique_types = thrust::count_if(
+      handle.get_thrust_policy(),
+      thrust::make_counting_iterator(size_t{0}),
+      thrust::make_counting_iterator(std::get<0>(unique_pairs_buffer).size()),
+      detail::is_first_in_run_t<edge_type_t const*>{std::get<0>(unique_pairs_buffer).data()});
+
+    unique_types.resize(static_cast<size_t>(nr_unique_types), handle.get_stream());
+    unique_type_counts.resize(static_cast<size_t>(nr_unique_types), handle.get_stream());
+
+    thrust::reduce_by_key(handle.get_thrust_policy(),
+                          std::get<0>(unique_pairs_buffer).begin(),
+                          std::get<0>(unique_pairs_buffer).end(),
+                          unique_pair_counts.begin(),
+                          unique_types.begin(),
+                          unique_type_counts.begin());
```

Comment:
Same here, copy_if + binary_search will be faster.

Reply:
For a type t, we need to sum up all the counts for type t. We need to touch each element in any case.
And, this :reduce_by_key runs on (type, count) pairs, not on entire list.